### PR TITLE
Sql tutorial - Lessons 1 to 6

### PR DESCRIPTION
**Pull request by jmbriano**
This PR is to extend codebar tutorials section with an SQL tutorial.

These first 6 lessons are the result of the collaboration of (at least) the following people to whom I'm very thankful for helping me pushing this initiative forward. In chronological order of collaboration: @brunogirin @daverick @kodikos @anufern @MarckK and @leahknight32

Should we also include in this PR the changes in the main pages to add this new section, please let me know how/where to do it and I'll extend the PR.

I'm looking forward to hear your feedback

### CODE REVIEW

File: sql/lesson1/tutorial.md [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+Open this fiddle that will have the data shown in the tables above already loaded: https://sqliteonline.com/#fiddle-5be8825d749e251ljod9s39t
+
+* L1.1 Show all the data available of all the people.
+* L1.2 Show the name and Country of residence of all the people 23 years old or above.
+* L1.3 Show the name and age of all the people living in UK.
+* L1.4 Show the name and age of all the people not living in their birth place.
+* L1.5 Show the distinct countries of residence of all the people.
+* L1.6 Show the distinct studies of the people having less than 2 children.
+
+We recommend you to try the above exercises in https://sqliteonline.com, if you get blocked, you can find the solutions [here](solutions.md)
```
Comment (kodikos) [inline sql/lesson1/tutorial.md:232]: Recommend replacing `blocked` with `stuck`

File: sql/lesson1/tutorial.md [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+
+## About Lesson 1
+
+_Lesson 1_ will be an introduction to SQL and its querying possibilities. You'll work with an online database already filled in with some data and you'll focus on the **SELECT** operator. But before doing so, you'll learn what a **TABLE** is and how it is structured. At the end, you'll find some exercises to _capture_ the knowledge.
+
+## What is a SQL?
+
+SQL stands for Structured Query Language. As explained above, it is a programming language for interacting with databases. SQL commands can be classified under: data structure; data manipulation; and data querying.
+
+Unlike most programming language it's a declarative language not an imperative one: you describe what you want as a result and you let the computer decide how to do it.
```
Comment (kodikos) [inline sql/lesson1/tutorial.md]: The first use of languages here should be plural

File: sql/lesson1/tutorial.md [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+## Style guide
+
+During these tutorials you will find SQL code fragments. There is a lot of bibliography out there about naming conventions and formatting recommendations, but it is not always possible to follow all of them. Thus, we have picked one style for these lessons. If you want to have a deeper look, you can find it [here](https://www.sqlstyle.guide/).
+
+## About Lesson 1
+
+_Lesson 1_ will be an introduction to SQL and its querying possibilities. You'll work with an online database already filled in with some data and you'll focus on the **SELECT** operator. But before doing so, you'll learn what a **TABLE** is and how it is structured. At the end, you'll find some exercises to _capture_ the knowledge.
+
+## What is a SQL?
```
Comment (kodikos) [inline sql/lesson1/tutorial.md]: Is the `a` intended?

File: sql/lesson2/tutorial.md [hunk 4]
```diff
@@ -0,0 +1,616 @@
+---
+layout: page
+title: Introduction to SQL
+---
+
+## About Lesson 2
+
+_Lesson 2_ will extend your knowledge of the **SELECT** operator by explaining some new operators, ways to sort the results and to **LIMIT** the number of results shown. You'll continue working with an online database already filled in with some data. At the end, you'll find some exercises constructing on top of the lessons 1 and 2.
```
Comment (kodikos) [inline sql/lesson2/tutorial.md]: Suggest replacing `constructing on top of the` to `building on`

File: sql/lesson2/tutorial.md [hunk 5]
```diff
@@ -0,0 +1,616 @@
...
+
+### Ordering query results
+
+Normally, when you write a query, you'll want your results to be shown in a specific order. Doing so is very simple knowing the operator **ORDER BY**.
```
Comment (kodikos) [inline sql/lesson2/tutorial.md]: Suggest `Doing so is very simple knowing the operator ***ORDER BY***` should be `You can achieve this using the **ORDER BY*** operator`

File: sql/lesson2/tutorial.md [hunk 6]
```diff
@@ -0,0 +1,616 @@
...
+* **{direction}** is either ASC or DESC, if omitted ASC is assumed.
+
+Using the same database as in Lesson 1 (https://sqliteonline.com/#fiddle-5be8825d749e251ljod9s39t), run the following query.
+```SQL
+SELECT name, age
+  FROM people
+ ORDER BY age DESC;
+```
+
+Is the result sorted by descending age? What happen in ages that appear more than once? Are the names sorted?
```
Comment (kodikos) [inline sql/lesson2/tutorial.md]: I think `what happen in ages` should be `what happens for ages`?

File: sql/lesson2/tutorial.md [hunk 7]
```diff
@@ -0,0 +1,616 @@
...
+```SQL
+SELECT name, age
+  FROM people
+ ORDER BY age DESC
+ LIMIT 5;
+```
+
+Contrary to the query used before without LIMIT that returned all the rows in the table, this one only returns the first 5.
+
+Although it is not strictly necessary to have an ORDER BY clause when you LIMIT the results, it is highly recommended. Doing so, you can be certain that you will always get the same results when running the same query multiple times.
```
Comment (kodikos) [inline sql/lesson2/tutorial.md]: I'm not sure about this para.. I can't think of a situation that ORDER BY becomes optional just because a LIMIT is used? Unless there is some assumption made about the natural ordering of the table (or a VIEW, but we haven't got to those yet)

File: sql/lesson2/tutorial.md [hunk 8]
```diff
@@ -0,0 +1,616 @@
...
+
+If additionally you are paginating the result, you need to use LIMIT together with **OFFSET**. This operator can we understood as _Skip the first N records, and give me the ones after_. For example, if you run
```
Comment (kodikos) [inline sql/lesson2/tutorial.md]: Maybe you should start this para more like `If you wanted to view the next _x_ results, you can specify an ***OFFSET***`? Also, I have never used OFFSET in my entire SQL career, I always use the shortened `LIMIT <count>,<offset>` format. Happy to conceded that's a MySQL thing, but just wondered if it was also worth mentioning here? Or is this a coaches easter egg? ;)

File: sql/lesson2/tutorial.md [hunk 9]
```diff
@@ -0,0 +1,616 @@
...
+ WHERE studies IS NOT NULL;
+```
+
+Study the results of each one. Easy huh?
+
+It's time to do some exercises now... :)
+
+## Lesson 2 exercises
+
+For the exercises below, we'll use the table from previous lesson: **people** and we'll introduce four new ones: **locations**, **hosts**, **workshops** and **attendance**. The next sections show a few rows as examples of the values. If you want to see the whole table, you'll find the fiddle link and the SELECT statement under each table.
```
Comment (kodikos) [inline sql/lesson2/tutorial.md]: A few corrections, if I may:
`For the exercises below, we'll use the **people** table from the previous lesson, and we'll introduce four new tables: **locations**, **hosts**, **workshops** and **attendance**.`
I'm not sure about the value of listing the fields and sample data - we haven't taught them table defs or insert statements, so we should probably just be pointing them straight to the fiddle hashes at the bottom of the tutorial?

File: sql/lesson2/tutorial.md [hunk 10]
```diff
@@ -0,0 +1,616 @@
...
+* L2.5 Show the location of cities in UK.
+* L2.6 Show all the host names.
+* L2.7 Show the host name and the student capacity. Sort the result in capacity descending order.
+* L2.8 Pick a host ID and show the dates where workshops were held there. Order the result chronologically.
+* L2.9 Show the top 5 host according to their total capacity (Students + Coaches)
+* L2.10 Show the distinct ID of the people that RSVP but then they did not show up (false = 0, both indicate individual did not attend the event).
+* L2.11 Show dates when workshops were held in these 3 hosts: _Edincode_, _Kanguland_, and _CodeLab_.
+    To do so, first find the host IDs. Then use those IDs to show the dates when workshops were held at those hosts (TIP: think of the usage of **IN**)
+
+We recommend you to try the above exercises in https://sqliteonline.com, if you get blocked, you can find the solutions [here](solutions.md)
```
Comment (kodikos) [inline sql/lesson2/tutorial.md]: As before, suggest replacing `blocked` with `stuck`

File: sql/lesson3/tutorial.md [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+#### Combining them
+
+One way of combining them is by adding more than one function in the SELECT statement. For example:
+```SQL
+SELECT MAX(age), MIN(age),
+       MAX(number_of_children), SUM(number_of_children)
+  FROM people;
+```
+
+### Aliases
```
Comment (kodikos) [inline sql/lesson3/tutorial.md:80]: Worth showing `AS` syntax too? I think it's technically more standard.

Comment (kodikos) [reply sql/lesson3/tutorial.md:80]: Also, this is the first place where you start to refer to more complicated field names - worth mentioning backticks?

File: sql/lesson5/tutorial.md [hunk 12]
```diff
@@ -0,0 +1,181 @@
+---
+layout: page
+title: Introduction to SQL
+---
+
+## About Lesson 5
+
+The previous lessons showed you how to manipulate data from a single table. _Lesson 5_ will show you how to combine data from multiple tables and will bring out the _relational_ part of relational databases.
+
+Tables in a relational database relate to each other as they contain keys which allow information from multiple tables to be combined easily. The real power of relational databases, and SQL, comes from working with multiple tables at once.
```
Comment (kodikos) [inline sql/lesson5/tutorial.md]: May I suggest a slight alteration, to use existing terminology and explain the concept a little clearer? Something along the lines of:
`Tables in a relational database are related by having common values in the fields of the tables. These fields are known as keys. These allow information from multiple tables to be combined...`

File: sql/lesson5/tutorial.md [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+
+In this lesson we will keep using the same database as before (https://sqliteonline.com/#fiddle-5be888fa749e451ljodase3y)
+
+### Primary and Foreign Keys
+
+All tables in a relational database need to have a primary key. This is an identifier that uniquely identifies each row. We saw in _Lesson 4_ that the `people` table uses the `id` field as a primary key. The key aspects of the primary key field are that it is never empty and it is unique: every record in a table has one and only one record can have a given primary key. For instance, in the `people` table, record with `id` `1` is the record for `Emma Smith`. If we try to insert a new record with the same `id` primary key, it will fail. If we try to insert a record with no value for `id` it will fail too.
+
+Any table can also have a number of foreign key fields. Each foreign key contains the value for a primary key in another table. For example, if we select the data in the `hosts` table:
```
Comment (kodikos) [inline sql/lesson5/tutorial.md]: Suggest improvement along the lines of:
`A foreign key is a field that contains the primary key value of another table, to indicate which record from that table relates to this row. This is known as a _reference_.`
Suggest concept of multiple foreign keys per table comes later.

File: sql/lesson5/tutorial.md [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+```SQL
+SELECT *
+  FROM locations;
+```
+
+This shows us that the record in the `locations` table that has `id` set to `2` is the location record for Brighton in the UK. That's great, we can now go from `hosts` to `locations` with two queries but can we get this data in a single query and for all records? This is where table joins come in.
+
+### Inner join
+
+#### Joining two tables
```
Comment (kodikos) [inline sql/lesson5/tutorial.md:36]: Suggest in this section you dissect the statement syntax to make it clear what the ON part is about, whereupon you can refer to it as specifying the _reference_ (primary key and foreign key), then it ties up a bit better with the explanation above.

Comment (kodikos) [reply sql/lesson5/tutorial.md:36]: This may also help them to disambiguate the syntax with a WHERE clause (and we'll just keep quiet about how it essentially _IS_ a where clause :D )

File: sql/lesson5/tutorial.md [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+SELECT hosts.name, locations.city, locations.country
+  FROM hosts
+  JOIN locations
+    ON hosts.location_id = locations.id
+ WHERE locations.country = 'UK';
+```
+
+#### Joining more than two tables
+
+You can have multiple **JOIN** clauses when connecting more than two tables. So for example if we wanted to know all workshop dates with host and location, we could do:
```
Comment (kodikos) [inline sql/lesson5/tutorial.md:95]: As explained above, suggest here is a good place to say that a table can have multiple foreign keys to join several tables' data to this one row.

File: sql/lesson4/solutions.md [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+
+## L4.8 Add a new location.
+
+Solution 1:
+```SQL
+INSERT INTO locations (id, city, country)
+VALUES (20, 'Seattle', 'US');
+```
+
+Solution 2:
```
Comment (kodikos) [inline sql/lesson4/solutions.md:119]: Don't think you've introduced subqueries at this point, so this may be a little OTT to add here :/

File: sql/lesson5/tutorial.md (hunk 14 above)
Comment (kodikos) [reply sql/lesson5/tutorial.md:36]: Also, this is where you introduce the `.` notation for fields in tables, so worth adding to the statement dissection.

File: sql/lesson6/tutorial.md [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+
+## About Lesson 6
+
+The previous lessons showed you how to manipulate data from a single table and from multiple tables using **JOIN**. _Lesson 6_ will show you how to use aliases, the **EXISTS** operator, and how to save complex queries into **VIEWs**.
+
+In this lesson we will keep using the same database than before (https://sqliteonline.com/#fiddle-5be888fa749e451ljodase3y)
+
+### Using aliases
+
+The moment we start combining tables as we did in _Lesson 5_, we need to tell the database engine from which table we want to pick the columns. So far, the `.` notation to explicitly specify the tables worked since we only joined a given table once per query. But, what happens if we need to join the same table more than once in the same query? Let's have a look at this example:
```
Comment (kodikos) [inline sql/lesson6/tutorial.md]: Sentence starting `So far` should be `So far, we've used`. Also, later in the sentence, suggest replacing `worked since` with `used, as`

File: sql/lesson6/tutorial.md [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+Sometimes you need to combine more than one table in a query to add more columns to the results, but in some cases, you only need other tables to check some conditions. Imagine you want to list all the persons that has RSVP to at least one event. You could do this using JOINs as follows:
+
+```SQL
+SELECT DISTINCT p.name
+  FROM people p
+       INNER JOIN rsvp event
+       ON event.person_id = p.id;
+```
+
+But if you look closer, we are not using any column from the _rsvp_ table in the result. In cases like this, we can use the **EXISTS** operator. Look at the query below, it produces the same result:
```
Comment (kodikos) [inline sql/lesson6/tutorial.md:55]: Isn't this a case for using `HAVING`? or is that only on GROUP BY's that don't refer to values in the result?

File: sql/lessons_summary.txt [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+    - Show the top 5 host according to their capacity
+    - Show the distinct ID of the persons that RSVP but then they did not show up (attendance = false).
+
+
+
+Lesson 3:
+- Functions in WHERE clause: MAX, MIN, COUNT, AVG
+- Functions in SELECT clause: UPPER,
+- GROUP BY
+- HAVING
```
Comment (kodikos) [inline sql/lessons_summary.txt]: This is not in tutorial 3. It's also mentioned in tutorial 7, but I only see tutorials 1-6. Is this page up to date?

File: sql/lessons_summary.txt [hunk 20]
```diff
@@ -0,0 +1,128 @@
+SQL student path
```
Comment (kodikos) [inline sql/lessons_summary.txt]: I don't quite get what this file is for? It's not referred to anywhere.. did you accidentally add your lesson plan to the PR? :D

File: sql/lesson1/tutorial.md (hunk 3 above)
Comment (jmbriano) [reply sql/lesson1/tutorial.md]: Just a typo. Will fix in next commit

File: sql/lesson1/tutorial.md (hunk 2 above)
Comment (jmbriano) [reply sql/lesson1/tutorial.md]: Fixed

File: sql/lesson1/tutorial.md (hunk 1 above)
Comment (jmbriano) [reply sql/lesson1/tutorial.md:232]: Replaced in the 6 lessons

File: sql/lesson2/tutorial.md (hunk 4 above)
Comment (jmbriano) [reply sql/lesson2/tutorial.md]: done

File: sql/lesson2/tutorial.md (hunk 5 above)
Comment (jmbriano) [reply sql/lesson2/tutorial.md]: done

File: sql/lesson2/tutorial.md (hunk 6 above)
Comment (jmbriano) [reply sql/lesson2/tutorial.md]: done

File: sql/lesson2/tutorial.md (hunk 7 above)
Comment (jmbriano) [reply sql/lesson2/tutorial.md]: I'll rephrase this to: `When using LIMIT you should always include an ORDER BY clause in your SQL statements. Doing so, you can be certain that you will get the same results when running the same query multiple times.`

File: sql/lesson2/tutorial.md (hunk 8 above)
Comment (jmbriano) [reply sql/lesson2/tutorial.md]: I learnt something :) 
I'm adding the following:

There is a short form when using LIMIT and OFFSET together. The following two queries produce the same results:

```SQL
SELECT name, age
  FROM people
 ORDER BY age DESC
 LIMIT 5
OFFSET 3;
```
```SQL
SELECT name, age
  FROM people
 ORDER BY age DESC
 LIMIT 3,5;
```

When using this short form, the first number corresponds to the _offset_ and the second one to the _number of records returned_.

File: sql/lesson2/tutorial.md (hunk 9 above)
Comment (jmbriano) [reply sql/lesson2/tutorial.md]: I think a visual representation of the tables can help the coaches and the students in the learning process. I suggest leaving them there and if we find them confusing we can remove them in a next iteration.

The first sentence was rephrased as suggested.

File: sql/lesson2/tutorial.md (hunk 10 above)
Comment (jmbriano) [reply sql/lesson2/tutorial.md]: done

File: sql/lesson3/tutorial.md (hunk 11 above)
Comment (jmbriano) [reply sql/lesson3/tutorial.md:80]: Added AS as indeed is the approach in the selected style guide. Backticks are not in the style guide so I propose to leave the queries as they are now. This can also avoid confusion with the quotes around the aliases names.

Do you agree?

File: sql/lesson4/solutions.md (hunk 16 above)
Comment (jmbriano) [reply sql/lesson4/solutions.md:119]: #goodcatch. I'll add this clarification:

`This solution satisfies the request, but it may be difficult to understand until you finish lesson 5. Skip it if it's not clear yet. You can always come back later to check it again.`

File: sql/lesson5/tutorial.md (hunk 12 above)
Comment (jmbriano) [reply sql/lesson5/tutorial.md]: Done

File: sql/lesson5/tutorial.md (hunk 13 above)
Comment (jmbriano) [reply sql/lesson5/tutorial.md]: Done

File: sql/lesson5/tutorial.md (hunk 14 above)
Comment (jmbriano) [reply sql/lesson5/tutorial.md:36]: I'll add the following dissection:

An INNER JOIN query can be generalized as follows:
```SQL
SELECT {columns}
  FROM {table_A}
 INNER JOIN {table_B}
    ON {table_B.primary_key} = {table_A.foreign_key_to_B}
```
In the queries above, the `.` is used to indicate to which table does the field correspond to. This is known as _dot notation_. When any of the fields in your statement can be found in more than one of the table involved, you must use this notation to make it clear to which one you are referring to.

File: sql/lesson5/tutorial.md (hunk 15 above)
Comment (jmbriano) [reply sql/lesson5/tutorial.md:95]: Good point. I'll add the following:

In the example above, **workshops** table has a foreign key (host_id) to **hosts** table, and **hosts** table has a foreign key (location_id) to **locations** table.

In some cases, you can find tables having more than one foreign key to other tables. For example, table **rsvp** has two foreign keys: one to **workshop** table (workshop_id) and one to **people** table (person_id).

File: sql/lesson6/tutorial.md (hunk 17 above)
Comment (jmbriano) [reply sql/lesson6/tutorial.md]: done

File: sql/lesson6/tutorial.md (hunk 18 above)
Comment (jmbriano) [reply sql/lesson6/tutorial.md:55]: I'm not sure I understood this question. HAVING is to be used only when using GROUP BY. And I don't see any GROUP BY around this comment so I'm a bit lost :(

File: sql/lessons_summary.txt (hunk 20 above)
Comment (jmbriano) [reply sql/lessons_summary.txt]: It should not be in the PR. It was meant to be a guide while writing and I accidentally added it to the PR.
I'll remove it

File: sql/lessons_summary.txt (hunk 19 above)
Comment (jmbriano) [reply sql/lessons_summary.txt]: I will remove this file, it's not updated and should not be in the PR either.

**Comment by jmbriano** [comment]
Hi @kodikos thanks **a lot** for the feedback. As you can see I implemented almost all of the comments. I only left 3 open for your review/clarification.

Let me know your thoughts when you have some time. And of course, if you want to comment in any implementation of the ones I closed please feel free to reopen or add new ones.

Martin

File: sql/lesson3/tutorial.md (hunk 11 above)
Comment (kodikos) [reply sql/lesson3/tutorial.md:80]: Yeah, backticks and single quotes could well be synonymous in use, actually, or MySQL-specific. If normal quotes work, that's fine. I didn't notice if you explicitly mentioned them, which might be worth doing, but not essential.

File: sql/lesson6/tutorial.md (hunk 18 above)
Comment (kodikos) [reply sql/lesson6/tutorial.md:55]: Fine, ignore me about this :)

**Comment by kodikos** [comment]
Well done on those fixes, it's looking really good :) Happy to help, but alas I've run out of time for now :( Also, not sure if I have write access to the repo to approve for you.

**Review by MarkCBJSS — COMMENTED** [review]
I've added this to PR438 - which also includes corrections to links for the `solutions.md` files within the tutorial pages and adds links from the tutorials `index.html` page.
